File: frontend/src/engine/capabilities/copyPasteRange.ts
/**
 * copyPasteRange – Copy a range and paste to another location.
 *
 * Supports: all, values only, formats only, formulas only.
 * Uses Range.copyFrom() which handles cross-sheet copies.
 */

import { CapabilityMeta, CopyPasteRangeParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";
import { resolveRange } from "./rangeUtils";

const meta: CapabilityMeta = {
  action: "copyPasteRange",
  description: "Copy a range and paste to another location",
  mutates: true,
  affectsFormatting: false,
};

async function handler(
  context: Excel.RequestContext,
  params: CopyPasteRangeParams,
  options: ExecutionOptions
): Promise<StepResult> {
  if (options.dryRun) {
    return { stepId: "", status: "success", message: `Would copy ${params.sourceRange} to ${params.destinationRange}` };
  }

  options.onProgress?.(`Copying ${params.sourceRange} → ${params.destinationRange}...`);

  const source = resolveRange(context, params.sourceRange);
  const dest = resolveRange(context, params.destinationRange);

  const pasteTypeMap: Record<string, Excel.RangeCopyType> = {
    all: "All" as Excel.RangeCopyType,
    values: "Values" as Excel.RangeCopyType,
    formats: "Formats" as Excel.RangeCopyType,
    formulas: "Formulas" as Excel.RangeCopyType,
  };

  const copyType = pasteTypeMap[params.pasteType ?? "all"] ?? ("All" as Excel.RangeCopyType);
  dest.copyFrom(source, copyType);
  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Copied ${params.sourceRange} → ${params.destinationRange} (${params.pasteType ?? "all"})`,
    outputs: { outputRange: params.destinationRange },
  };
}

registry.register(meta, handler as any);
export { meta };

File: backend/app/execution/capabilities/copy_paste_range.py
"""copyPasteRange — copy a source range to a destination range."""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry
from app.execution.range_utils import resolve_range


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    src = params.get("sourceRange") or params.get("source")
    dst = params.get("destinationRange") or params.get("destination")
    paste_type = (params.get("pasteType") or "all").lower()
    if not src or not dst:
        return {"status": "error", "message": "copyPasteRange requires 'sourceRange' and 'destinationRange'."}

    source = resolve_range(ctx.workbook_handle, src)
    destination = resolve_range(ctx.workbook_handle, dst)

    if ctx.dry_run:
        return {
            "status": "preview",
            "message": f"Would copy {source.address} → {destination.address} ({paste_type}).",
        }

    # xlwings Range.copy(destination) defaults to "all" (values + formats +
    # formulas). For values-only / formats-only we fall back to property
    # assignment.
    if paste_type in ("all", "allwithoutborders"):
        source.copy(destination)
    elif paste_type == "values":
        destination.value = source.value
    elif paste_type in ("formats", "format"):
        destination.number_format = source.number_format
        try:
            destination.color = source.color
        except Exception:
            pass
    elif paste_type == "formulas":
        destination.formula = source.formula
    else:
        return {"status": "error", "message": f"Unknown pasteType {paste_type!r}."}

    return {
        "status": "success",
        "message": f"Copied {source.address} → {destination.address} ({paste_type}).",
        "outputs": {"sourceRange": source.address, "destinationRange": destination.address},
    }


registry.register("copyPasteRange", handler, mutates=True)
